Options for handling symlinks during hfsplus addall

AdamNorberg opened the pull request:
Chromium includes symlinks inside its application bundle (separate from the link from ` ` to `/Applications` at the DMG root as part of the "drag install" UI). To use `libdmg-hfsplus` with Chromium, I need it to support symlinks by copying the link itself rather than traversing it.

This pull request implements this. It's built on top of pull requests #17, #18, #19, and #20; the commits unique to this sequence begin at `Use lstat to determine entity type`. (If Github pull requests have a good way to represent a chain of stacked changes, please let me know and I'll use it. I'm minimally familiar with it, I am used to other code review tools, sorry about that!)

The new behaviors are all guarded behind flags; behavior should be identical to the existing logic if they are not provided. `hfsplus addall` now takes `--symlinks=` or `-s`, with accepted values `fail`, `traverse`, and `copy`:

* on `fail`, the tool will exit with an error message if it finds a symlink in the input
* on `traverse`, the tool will follow symlinks and write the item at the destination into the raw UDIF as though the target file was literally placed at the symlink's destination; this is the default behavior because it is the behavior before the patch
* on `copy`, the symlink itself is recreated literally in the new location

Separately, this adds a new flag to control whether `hfsplus addall` should automatically assign special permissions and file modes to specific paths in the archive, if it finds itself to be making them. Comments, the history of the code, and the paths themselves suggest this is for building installation packages intended for jailbroken iPhones, or to deploy jailbreak payloads onto iPhones. Users that run into this by accident are likely to be unpleasantly surprised, so `--special-modes=no` or `-m0` or similar (see `must_parse_bool` in hfs.c for details of accepted values) will turn this behavior off. I decided to use an explicit value rather than a "presence or absence" flag because either the default would be equivalent to providing the flag (for backwards compatibility reasons), necessitating `--no-special-modes` to get it to do anything useful, or I have to implement the flag in a "backwards"-feeling way, where it is provided to _not_ do a special case extra behavior. Given a choice between two unpleasant options, I punted and decided to only take the flag explicitly. This is not a very GNU-utilities thing to do -- either of the other options would be more standard -- so I am very open to changing this if a maintainer has opinions.

Right now, at time of providing this pull request, it's not ready to go -- I need to write test automation. This is challenging because the compression libraries on my Mac don't produce bit-for-bit identical results to those expected  in your test code. Running regression tests, after this patch I get the same results I get from running at `tip` of the canonical version of the repository, down to the checksums --a result that says half the tests failed because those checksums were wrong, along with details of the diffs. The `diff` utility also appears to make different decisions regarding whether to break a run of consecutive changed lines in `--unified` mode for a single unchanged line (mine doesn't, your test expectations seem to anticipate it). There's no mystery to this one -- your tests are designed to run under Ubuntu Linux (the primary target for libdmg-hfsplus -- we're planning to run it on a Mac for vaguely complicated reasons), not macOS 15. I'll get started on using the Docker-based setup and see where I get with that; VMs are discouraged on dev boxes here so I have some extra roadblocks to work through first.

AdamNorberg committed:
First bounds check on fullName.

AdamNorberg committed:
Wrap "check and cast" pattern for HFSPlusCatalogFolder.

If we use this literally everywhere, it might make it easier to convert to a union-based approach that does not violate the strict aliasing rule. That is probably unreasonably ambitious. However, I would rather have these helpers available instead of repeating the check everywhere, and am likely to rewrite the pattern where I see it, especially if I see repeated casts of the same (ambiguous) pointer, so here it is for my personal convenience.

AdamNorberg committed:
Add bounds check to appending items in addAllInFolder. Fix hfslib.c include syntax.

AdamNorberg committed:
Clean up addAllInFolder overwrite case.

This uses the "check and cast" helpers and adds error checks to handle
unexpected record subtypes. The result is longer but more legible; I
feel like the original was past the line of what is reasonable to
do with the selection operator.

AdamNorberg committed:
Use lstat to determine entity type.

AdamNorberg committed:
Add flags for symlink traversal and special chmods.

AdamNorberg committed:
addall_hfs_2: accept symlink and special-case file dispositions

AdamNorberg committed:
fix my many mistakes, refactor hfs.c error handling

AdamNorberg committed:
Actually use the flag

AdamNorberg committed:
add debug message to addAllInFolder2

AdamNorberg committed:
NUL-terminate readlink's output. Tricksy little hazard there, readlink.

AdamNorberg committed:
Clean up a smattering of memory leaks and missing error checks.

AdamNorberg committed:
Oops, I damaged a variable name

AdamNorberg commented:
Regression tests pass cleanly when using your Docker setup (with Podman) on a Linux host. With that working, I'll develop new tests for symlink scenarios now.

AdamNorberg commented:
I've gotten distracted into trying to figure out if there's any way to get bit-perfect results out of _any_ test involving `hfsplus addall` (I am tentatively concluding that, presently, there is no way to do so because timestamps are always set to "now", there is no plumbing to preserve them from the input, so I would need to freeze the clock in the VM, override `time` some other way, or create an additional command line flag to functionally override `time` with a single fixed timestamp. Might be worth pursuing but I probably should not put it on the critical path of this patch.) In so doing, I've noticed that symlink _extraction_ is special-cased in Windows -- see `extractOne`, in hfs/hfslib.c, look for `#ifdef WIN32` -- presumably because symbolic links are ostensibly not a win32 feature. [It turns out they now are](https://learn.microsoft.com/en-us/windows/win32/fileio/symbolic-links) but presumably that is newer than the 11-year-old patch ("create copies instead of symlinks on Windows") that implemented this.

Should I do anything about Windows here? Options that come to mind:
* no. symlinks work. hfs addall will handle them fine. they won't round-trip because of that special case, but that does not affect addall.
* block the "copy" disposition -- if a user attempts to use it in the Windows version, crash with a complaint that "hfsplus is older than symlinks on Windows and therefore handles them inconsistently, so let's not".
* Strip out the `#ifdef WIN32` special cases pertaining to symlinks. They are outdated.
* Create a new family of definable macros:
    * `FORBID_HOST_SYMLINKS`, which would block symlink behavior on all platforms
    *  `ALLOW_WIN32_SYMLINKS`, which would unblock symlinks on Windows
    * `HOST_SYMLINKS` (possibly with a different name so it is obvious that this is not to be specified with `-D` or similar), which a common `macros.h` would define as functionally equivalent to `(ALLOW_WIN32_SYMLINKS || !WIN32) && !FORBID_HOST_SYMLINKS` (if these were bools rather than defined-or-undefined, but you get the idea), which actually guards symlink behavior and the `copy` disposition

Having typed all this out, I am leaning towards "do not do anything about this right now, but it might be nice to revisit all `#ifdef WIN32`s at some nebulous future time", because it's not going to make existing behavior any worse, if a Windows user wants to `copy` symlinks they probably expect that to work and there is no reason it shouldn't, and cleaning up the unnecessary `win32` flags is out of scope for this patchset.

AdamNorberg committed:
Add prototype of functional test for symlink handling.

AdamNorberg committed:
Switch from strlcpy to strlen + memcpy.

strlcpy, while convenient, is a BSD API not available on the Debian image used for the CI workflow (although the Ubuntu image specified in the Dockerfile is perfectly fine with it). Since strncpy does not provide any way to distinguish between "okay, I copied your entire string" and "btw I couldn't find room for the NUL terminator, hope you didn't need that, lul", I'm using strlen+memcpy explicitly, which is at least more obvious about how it will not help you maintain C string validity in any way.

This implementation relies on the original string to have a \0 in the expected spot, but so does strlen (and all other str-based APIs) so it's not like this makes the situation worse.

AdamNorberg committed:
fix typo

AdamNorberg committed:
Merge branch 'scattered-substantiative-cleanups' into symlink-feature

AdamNorberg committed:
Merge branch 'symlink-feature' into symlink-feature-test

AdamNorberg committed:
cd instead of pushd, fix missing $

AdamNorberg committed:
oh, I don't need to use echo to check retval in cram. cool!

AdamNorberg commented:
Please take a look at `addall_symlink_modes.t` for functional testing of symlink handling.

Not covered: the "special modes for special files" flag. I remembered I need to test that too about two seconds after I clicked the "ready for review" button. My testing for that will be minimal; I'm not going to attempt to provoke/observe the setuid bit behaviors, because umask should make them difficult to observe, but I can look for the x bit behaviors on stuff that looks like `Applications/*.app`. I'll start on that tomorrow, should go much faster now that I know how to make Cram work and my test environment work.

AdamNorberg committed:
remove another strlcpy and prevent it from recurring

AdamNorberg commented:
Oops, I added another strlcpy in this set of changes, which i missed when I searched for `strlcpy` previously because I did that at the branch I was trying to fix and forgot to redo it after merging that fix into this branch.

Because I keep making this mistake, I will hubristically assume that other hypothetical future maintainers might also make the same mistake while Firefox's build is using a libc without it, so I have added a few lines to `common.h` as a speed bump. Is this too much? I can take it back out.

AdamNorberg committed:
Prototype a test for path-triggered chmod behaviors

AdamNorberg committed:
Fix a few wrong filenames

AdamNorberg committed:
oops, missed a spot

AdamNorberg commented:
This PR is now comprehensively tested and tests pass on my VM. (This includes a lot of new coverage for existing behaviors of `hfs addall`, including an apparent bug. I am choosing not to fix the apparent bug, because my sole interest in tinkering with the special path chmod behavior is adding a way to shut it off.) Please let me know if this breaks Mozilla's trybots so I can fix it -- and, of course, I would like to hear any code review feedback you have!

AdamNorberg committed:
Rename symlink policy "copy" to "clone_link".

When writing the tests, I kept getting myself confused because I kept expecting "copy" to make a copy of the target file, which is the exact opposite of what it does (it copies the link itself instead). If I get confused using it, and I'm the one who *wrote* it, it's almost certainly going to be confusing to everyone who tries to use it in the future, too. `clone_link` is much less ambiguous about what is happening.

I considered `link` but I worried it might make people think that it would create a link to the same destination, which is not quite what happens -- it creates a literally identical link, which means a relative link will be relative to its new location.  This is the best name I can think of for intuitively matching these semantics.

Separately, I am not thrilled with the flag name `--special_modes` but I keep not coming up with any better ideas.

AdamNorberg commented:
This is starting to land in Chromium (finally) and code review cleaned up a lot of things. If Mozilla folks are interested in this patch at all, I will merge my Chromium post-code-review changes back into this branch (of course I want to do that, I'd love to have less divergence from the upstream). I can't spend time keeping this PR/branch maintained if nobody will ever review it for a merge, though.

gabrielesvelto commented:
> This is starting to land in Chromium (finally) and code review cleaned up a lot of things. If Mozilla folks are interested in this patch at all, I will merge my Chromium post-code-review changes back into this branch (of course I want to do that, I'd love to have less divergence from the upstream). I can't spend time keeping this PR/branch maintained if nobody will ever review it for a merge, though.

Yes please, I apologize for the delay, I'll make sure I review your new changes before the end of the week.

AdamNorberg committed:
Merge remote-tracking branch 'origin/mozilla' into scattered-substantiative-cleanups

AdamNorberg committed:
Reindent abstractfile.c

AdamNorberg committed:
base64.c needs File, not dmg. switch the include.

This way, "common" can compile as an independent subcomponent (as I've specified Chromium's libdmg-hfsplus build to work); otherwise, the circular dependency requires "dmg" and "common" to always be compiled together.

AdamNorberg committed:
Reindent attribution.c.

AdamNorberg committed:
Reindent checksum.c a little.

AdamNorberg committed:
Reindent dmgfile.c (preferring tabs).

Tabs were predominant in this file,.

AdamNorberg committed:
Reindent dmglib.c (with tabs).

Tabs were predominant in this file.

AdamNorberg committed:
Reindent filevault.c. Use braces for one-line ifs.

Let's avoid a `goto fail` of our own.

AdamNorberg committed:
Reindent partition.c, clean up braces.

AdamNorberg committed:
Reindent and rebrace resources.c.

There were substantiative indentation errors in this file. I believe the indentation now matches the behavior.

AdamNorberg committed:
Fix uninitialized read in debugBTree.

AdamNorberg committed:
Fix indentation and braces in catalog.c.

AdamNorberg committed:
Reindent and rebrace volume.c.

AdamNorberg committed:
xattr.c: clean up updateAttributes types, add more braces.

AdamNorberg committed:
Fix indentation and trailing whitespace in dmg.h.

AdamNorberg committed:
Fix hfsplus.h indentation. Delete tryCatalogRecordAsThread family.

Nothing handles those cases anyway, the helper is just clutter. It can easily be added later if needed. (also the underlying file got removed in a catalog.c cleanup that was a paste from a version of the repo where I didn't add the extraneous helpers in the first place, oops)

AdamNorberg committed:
Back out from MAXPATHLEN, it was a bad idea.

The path length limit is an HFS+ thing, not a local system thing, so picking up MAXPATHLEN from POSIX libraries is a mistake; I'd need to add a constant in hfsplib.h instead. But if I'm doing that, I _really_ need to make sure it is consistent through the entire codebase, and I can't justify spending the time on that cleanup right now. So the bounds checks stay, but they will have to rely on length literals.

AdamNorberg committed:
Code review feedback -- assert cleanups, structure cleanups

AdamNorberg committed:
Merge branch 'scattered-substantiative-cleanups' into symlink-feature

AdamNorberg committed:
Cleanups from Chromium review.

Also de-MAXPATH-ifies the logic here too, fixing an unnecessarily
defensive (by one byte) range check along the way.

AdamNorberg commented:
This has picked up changes from the code review for PR #20, plus changes that came after code review from my Chromium teammates; it's ready for review. Thanks for picking these up again!

AdamNorberg commented:
@gabrielesvelto , I'm going to go on vacation for the rest of the year at the end of this week. If you can get some code review feedback to me in the next 24 hours, I should be able to get another iteration up before I disappear for a couple of weeks; otherwise, if substantial changes are required, I'm not likely to dig into them until the second week of January.

gabrielesvelto commented:
> @gabrielesvelto , I'm going to go on vacation for the rest of the year at the end of this week. If you can get some code review feedback to me in the next 24 hours

I'll see what I can do, worst case I'll finish the reviews before the end of the year so when you come back you won't have to wait on my feedback again.

gabrielesvelto commented on includes/hfs/hfslib.h:
```diff
@@ -11,13 +32,35 @@ extern "C" {
+	   symlinks in the input are handled, and `assignSpecialPermissions` controls
+	   whether to assign alternate permissions to some specific paths associated
+	   with system image installers, BootNeuter, and/or system binaries. */
+	void addAllInFolder2(
```
Calling this `addAllInfoFolderWithPolicies()` would make the intent more clear.

AdamNorberg replied:
Renamed to `addAllInFolderWithPolicies` (I assume `Info` was a typo).

gabrielesvelto commented on hfs/hfslib.c:
```diff
@@ -522,54 +552,51 @@ void extractAllInFolder(HFSCatalogNodeID folderID, Volume* volume) {
+	addall_hfs_2(volume, dirToMerge, dest, kIncomingSymlinksTraverse, TRUE);
+}
+
+void addall_hfs_2(
```
Same as with `addAllInFolder2()`, giving this a more descriptive name such as `addall_hfs_with_policies()`, or something along the lines, would make the code more readable.

AdamNorberg replied:
Done.

gabrielesvelto commented on hfs/hfslib.c:
```diff
@@ -522,54 +552,51 @@ void extractAllInFolder(HFSCatalogNodeID folderID, Volume* volume) {
-	
-	ASSERT(getcwd(cwd, 1024) != NULL, "cannot get current working directory");
-	
+	ASSERT(getcwd(cwd, 1024) != NULL, "addall_hfs_2: getcwd failed");
```
It would be nice to use `__func__` here and in all other asserts that have the function name. We could add it to `ASSERT()` everywhere, I think it would be a net positive for error messages. Something like:
```c
#define ASSERT(x, m) if(!(x)) { assertPrint(__func__, m); }

static inline void assertPrint(const char *func_name, const char *msg) {
  // ...
}
```
WDYT?

AdamNorberg replied:
Might as well go for `__func__`, `__FILE__`, `__LINE__` while we're at it!

gabrielesvelto reviewed (changes requested):
Just a couple of changes around naming and error messages, the rest LGTM.

AdamNorberg committed:
Implement code review feedback from gabrielesvelto.

AdamNorberg committed:
Fix test broken by change in ASSERT.

Helpfully, this shows the ASSERT patch works!

AdamNorberg commented:
Not a big fan of the KeBugCheck/KeBugCheckEx naming, I infer. 😁  hfsplus.h/catalog.c already has this going on (getRecordFromPath/2/3) so I stuck with it, but it is _not_ hard to convince me to pick names that explain themselves a little more.

gabrielesvelto reviewed (approved):
LGTM, thank you!
